### PR TITLE
Update mergesort.py

### PR DESCRIPTION
**Pull request by akashgkrishnan**
1) Updating the merge sort in python as the previous implementation was modifying the input array
2) divided the division part and conquer part of the merge sort algorithm as 2 functions namely mergeSort and merge.
3) function mergeSort divides the function into halves i.e the purpose of the function will be to divide the array
4) function merge will merge 2 halves into a sorted array
5)Added random test cases using shuffle as suggested by @dhruvmanila
6 The time and space complexity of the previous and my version remains the same. i.e (n log(n) time and n log(n) space
7) changed variables as per the python case as required and suggested by @dhruvmanila
8) Updated function names as suggested by @dhurvmanila

### **Describe your change:**

* [ ] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [ ] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: divide_and_conquer/mergesort.py [hunk 1]
```diff
@@ -1,48 +1,58 @@
...
+    >>> array = [-2, 3, -10, 11, 99, 100000, 100, -200]
+    >>> shuffle(array)
+    >>> merge_sort(array)
+    [-200, -10, -2, 3, 11, 99, 100, 100000]
+    >>> shuffle(array)
+    >>> merge_sort(array)
+    [-200, -10, -2, 3, 11, 99, 100, 100000]
+    >>> shuffle(array)
+    >>> merge_sort(array)
+    [-200, -10, -2, 3, 11, 99, 100, 100000]
```
Comment (dhruvmanila) [inline divide_and_conquer/mergesort.py:42]: Please use multiple test cases, some of the examples could be:
- Empty list
- One element list
- List containing strings, integers and other lists
- Mixed element types like a list containing integers, strings, etc
- Use `from string import letters`, `range`
- Verify the tests with built-in `sort` method or `sorted` function (easier than typing it out)

Be creative and mix it up in coming up with the test cases 😁

**Comment by dhruvmanila** [comment]
Thank you for your contribution!
The way pull request work can be tricky to understand but here's the gist of it:
- Fork the repo
- Make a separate branch for whatever you want to contribute
- Commit the changes in that branch
- Make the pull request using that branch
- To make further changes, push more commits to that branch and it will automatically show up here
- After your changes are merged, delete your branch or the repo (your choice)
- Repeat

File: divide_and_conquer/mergesort.py [hunk 2]
```diff
@@ -1,48 +1,58 @@
...
-    right_arr = arr[mid + 1 : right + 1]
-    k = left
-    i = 0
-    j = 0
-    while i < len(left_arr) and j < len(right_arr):
-        # change sign for Descending order
-        if left_arr[i] < right_arr[j]:
-            arr[k] = left_arr[i]
-            i += 1
+def merge(left_half, right_half):
```
Comment (dhruvmanila) [inline divide_and_conquer/mergesort.py:1]: Please use type hints for all functions and parameters.

Refer: https://www.python.org/dev/peps/pep-0484/
Or: https://docs.python.org/3/library/typing.html

Comment (akashgkrishnan) [reply divide_and_conquer/mergesort.py]: This was new to me. I did not know about type hints... Thanks for the suggestion

**Comment by akashgkrishnan** [comment]
@dhruvmanila kindly review the changes

**Comment by akashgkrishnan** [comment]
@dhruvmanila I have added in more test cases. Please let me know in case further improvements required

File: divide_and_conquer/mergesort.py [hunk 3]
```diff
@@ -1,48 +1,78 @@
...
-    j = 0
-    while i < len(left_arr) and j < len(right_arr):
-        # change sign for Descending order
-        if left_arr[i] < right_arr[j]:
-            arr[k] = left_arr[i]
-            i += 1
+from typing import List
+
+
+def merge(left_half: List, right_half: List) -> List:
```
Comment (dhruvmanila) [inline divide_and_conquer/mergesort.py:4]: Please provide `doctest` for this function as well.

**Comment by akashgkrishnan** [comment]
@dhruvmanila I have made the suggested changes and added in the doc test for the merge function as well.

**Review by dhruvmanila — APPROVED** [review]

**Pull request merged**